[Test] Use Ray client in XGBoost train_small release test

mwtian opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

This is an example of using Anyscale connect for a Ray release test. The script is tested against latest release of Ray, Anyscale CLI, and releaser from head. The command to run is the same as running other release tests:
```
$ python releaser/e2e.py --test-config ray/release/xgboost_tests/xgboost_tests.yaml --test-name train_small
```

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

AmeerHajAli commented on release/xgboost_tests/app_config.yaml:
```diff
@@ -1,4 +1,4 @@
-base_image: "anyscale/ray:1.2.0"
+base_image: "anyscale/ray-ml:pinned-nightly-py38"
```
you can also use :1.4.0 if you want.

mwtian replied:
Sounds good. I think if there is no other blocker, the nightly image may include newer versions of non-Ray dependencies, which could be more suitable for nightly / release tests.

AmeerHajAli commented on release/xgboost_tests/workloads/train_small.py:
```diff
@@ -9,14 +9,33 @@
+    # https://github.com/anyscale/product/issues/4931
+    compute_env_path = os.path.join(dir, "../app_config.yaml")
+    with open(compute_env_path, "r") as cluster_env:
+        builder.cluster_env(yaml.load(cluster_env, Loader=yaml.FullLoader))
```
alternatively, you can use the anyscale SDK to create a cluster env (look at the test_anyscale_development with connect) and reuse this cluster env in all the tests. This will significantly reduce the time it takes to run some of these tests.

krfricke replied:
+1, we have the functionality in the e2e script already, and it will solve a couple of problems, e.g. rendering the yaml jinja2 template and also creating the test data (which is created in the docker image).

The name to the cluster env could then be passed via a command line argument

mwtian replied:
Reusing the cluster created in the e2e script now.

ijrsvt reviewed (commented):
This overall looks pretty good. Let's sort out 4931 soon!

krfricke commented:
Re: 

```
xgboost_tests $ python create_test_data.py ./data/classification.parquet  # no sure if this is right
```

The problem here is that this command needs to be run on all nodes (head node and worker nodes). This is because all nodes should have access to the same data, and for testing purposes it is easier to just store them on a local disk rather than on a NFS or S3.

Previously we ran this command in the startup commands, so the dataset was created and written on all nodes after they booted. We moved this to the app config in the current iteration, which makes much more sense as the data will be stored on a separate layer in the docker container. Some other tests are using larger test data sizes, and this saves a lot of time. 

IMO here we should also aim to include this in the app config. I'd strongly vote to keep creating the app config with the e2e script and then just passing the identifier to `connect()` (if this is a valid API call).

mwtian commented on release/xgboost_tests/workloads/train_small.py:
```diff
@@ -9,14 +9,33 @@
+    builder = ray.client("anyscale://train_small").cloud(
+        "anyscale_default_cloud")
+
+    dir = os.path.dirname(os.path.realpath(__file__))
```
This logic is deleted now.

mwtian commented:
> Re:
> 
> ```
> xgboost_tests $ python create_test_data.py ./data/classification.parquet  # no sure if this is right
> ```
> 
> The problem here is that this command needs to be run on all nodes (head node and worker nodes). This is because all nodes should have access to the same data, and for testing purposes it is easier to just store them on a local disk rather than on a NFS or S3.
> 
> Previously we ran this command in the startup commands, so the dataset was created and written on all nodes after they booted. We moved this to the app config in the current iteration, which makes much more sense as the data will be stored on a separate layer in the docker container. Some other tests are using larger test data sizes, and this saves a lot of time.
> 
> IMO here we should also aim to include this in the app config. I'd strongly vote to keep creating the app config with the e2e script and then just passing the identifier to `connect()` (if this is a valid API call).



> Re:
> 
> ```
> xgboost_tests $ python create_test_data.py ./data/classification.parquet  # no sure if this is right
> ```
> 
> The problem here is that this command needs to be run on all nodes (head node and worker nodes). This is because all nodes should have access to the same data, and for testing purposes it is easier to just store them on a local disk rather than on a NFS or S3.
> 
> Previously we ran this command in the startup commands, so the dataset was created and written on all nodes after they booted. We moved this to the app config in the current iteration, which makes much more sense as the data will be stored on a separate layer in the docker container. Some other tests are using larger test data sizes, and this saves a lot of time.
> 
> IMO here we should also aim to include this in the app config. I'd strongly vote to keep creating the app config with the e2e script and then just passing the identifier to `connect()` (if this is a valid API call).

Thanks Kai for the explanations. With project/releaser#11, generating the data would be done in app configs and this command is no longer necessary.

krfricke commented:
Thanks - I'll take a look!

Please do not merge yet though, I want to make sure everything works with the current infrastructure. I'm happy to merge once it passes.

mwtian commented:
> Thanks - I'll take a look!
> 
> Please do not merge yet though, I want to make sure everything works with the current infrastructure. I'm happy to merge once it passes.

I see. I was thinking of merging the releaser change first. Let me make some changes to make sure this works with the current releaser.

krfricke commented on release/xgboost_tests/xgboost_tests.yaml:
```diff
@@ -8,9 +8,10 @@
     timeout: 600
     prepare: python wait_cluster.py 4 600
-    script: python workloads/train_small.py
+    script: workloads/train_small.py
```
Doesn't this break execution as we don't have a shebang?

mwtian replied:
I was thinking of handling `script` different when run as client, by passing `script` as one arg among the list of `args` to `subprocess.Run()`. Reverted now.

krfricke commented on release/xgboost_tests/workloads/train_small.py:
```diff
@@ -16,7 +16,7 @@
 
 if __name__ == "__main__":
-    ray.init(address="auto")
+    ray.client().job_name(os.environ["RAY_JOB_NAME"]).connect()
```
`os.environ.get("RAY_JOB_NAME", "no_name")`?

mwtian replied:
Only set `job_name()` in client mode now, and default to `train_small`.

krfricke reviewed (commented):
just two questions

mwtian commented:
The remaining requirement to make this test work in client mode is to use an Anyscale CLI that includes https://github.com/anyscale/product/commit/25a367f10f8f7d2863cdff1eb0e9ae1fa0bc2a81.
